ext/ffi/fast_call.rs

```diff
@@ -0,0 +1,1409 @@
+
+      (_, param) => {
+        let size_original = param.size();
+        // 16 and 8 bit integers are 32 bit integers in v8
+        let size_trampl = max(size_original, 4);
+        // https://developer.apple.com/documentation/xcode/writing-arm64-code-for-apple-platforms:
+        // > Function arguments may consume slots on the stack that are not multiples of 8 bytes.
+        // (i.e. natural alignment instead of eightbyte alignment)
+        //
+        // TODO: V8 does not currently follow this Apple's policy, and instead aligns all arguments to 8 Byte boundaries.
+        // A decision needs to be taken:
+        // 1. leave it broken and wait for v8 to fix the bug
+        // 2. Adapt to v8 bug and follow its ABI instead of Apple's. When V8 fixes the implementation, we'll have to fix it here as well
+        let padding_trampl =
+          (size_trampl - self.offset_trampoline % size_trampl) % size_trampl;
+        let padding_callee =
+          (size_original - self.offset_callee % size_original) % size_original;
+        match param {
+          I(B) | U(B) => dynasm!(self.assmblr
+            ; .arch aarch64
+            ; ldr w8, [sp, self.offset_trampoline + padding_trampl]
+            ; strb w8, [sp, self.offset_callee + padding_callee]
+          ),
+          I(W) | U(W) => dynasm!(self.assmblr
+            ; .arch aarch64
+            ; ldr w8, [sp, self.offset_trampoline + padding_trampl]
+            ; strh w8, [sp, self.offset_callee + padding_callee]
+          ),
+          I(DW) | U(DW) => dynasm!(self.assmblr
+            ; .arch aarch64
+            ;  ldr w8, [sp, self.offset_trampoline + padding_trampl]
+            ; str w8, [sp, self.offset_callee + padding_callee]
+          ),
+          I(QW) | U(QW) => dynasm!(self.assmblr
+            ; .arch aarch64
+            ; ldr x8, [sp, self.offset_trampoline + padding_trampl]
+            ; str x8, [sp, self.offset_callee + padding_callee]
+          ),
+        }
+        self.offset_trampoline += padding_trampl + size_trampl;
+        self.offset_callee += padding_callee + size_original;
+      }
+    };
+  }
+
+  fn zero_first_arg(&mut self) {
+    dynasm!(self.assmblr
+      ; .arch aarch64
+      ; mov x0, xzr
+    );
+  }
+
+  fn tailcall(&mut self, ptr: *const c_void) {
+    // stack pointer is never modified and remains aligned
+    // frame pointer remains the one provided by the trampoline's caller (V8)
+
+    // Like all ARM instructions, move instructions are 32bit long and can fit at most 16bit immediates.
+    // bigger immediates are loaded in multiple steps applying a left-shift modifier
+    let mut address = ptr as u64;
+    let mut imm16 = address & 0xFFFF;
+    dynasm!(self.assmblr
+      ; .arch aarch64
+      ; movz x8, imm16 as u32
+    );
+    address >>= 16;
+    let mut shift = 16;
+    while address > 0 {
+      imm16 = address & 0xFFFF;
+      dynasm!(self.assmblr
+        ; .arch aarch64
+        ; movk x8, imm16 as u32, lsl shift
```

Comment:
The body of the loop could be checking of the imm16 is zero.

For example, given `0x0000_XXXX_YYYY_0000`, the first instruction would emit a constant zero move.

Reply:
I'll add it. This is as unlikely as trivial is the additional logic. Thanks!